File: java/src/test/java/it/giannibombelli/wsc2026/booking/application/commands/BookingCommandsTest.java
package it.giannibombelli.wsc2026.booking.application.commands;

import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.primitive.GiftCardReference;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

class BookingCommandsTest {

    @Nested
    class PlaceBookingValidation {
        @Test
        void shouldFailIfParametersAreNull() {
            BookingId bookingId = EntityId.generate(BookingId::new);
            Money amount = new Money(BigDecimal.TEN);
            GiftCardReference giftCardReference = new GiftCardReference(UUID.randomUUID());

            assertThatThrownBy(() -> new PlaceBooking(null, amount, new Description("desc"), giftCardReference))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new PlaceBooking(bookingId, null, new Description("desc"), giftCardReference))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new PlaceBooking(bookingId, amount, null, giftCardReference))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new PlaceBooking(bookingId, amount, new Description("desc"), null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class ConfirmBookingValidation {
        @Test
        void shouldFailIfParametersAreNull() {
            BookingId bookingId = EntityId.generate(BookingId::new);
            GiftCardReference giftCardReference = new GiftCardReference(UUID.randomUUID());
            Money amount = new Money(BigDecimal.TEN);

            assertThatThrownBy(() -> new BookingConfirmationCommands.ConfirmBooking(null, giftCardReference, amount))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new BookingConfirmationCommands.ConfirmBooking(bookingId, null, amount))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new BookingConfirmationCommands.ConfirmBooking(bookingId, giftCardReference, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class RejectBookingValidation {
        @Test
        void shouldFailIfParametersAreNull() {
            BookingId bookingId = EntityId.generate(BookingId::new);
            GiftCardReference giftCardReference = new GiftCardReference(UUID.randomUUID());
            Money amount = new Money(BigDecimal.TEN);

            assertThatThrownBy(() -> new BookingConfirmationCommands.RejectBooking(null, giftCardReference, amount))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new BookingConfirmationCommands.RejectBooking(bookingId, null, amount))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new BookingConfirmationCommands.RejectBooking(bookingId, giftCardReference, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }
}

File: typescript/tests/booking/application/commands/bookingCommands.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import { placeBooking } from '@/booking/application/commands/placeBooking.js';
import {
  confirmBooking,
  rejectBooking,
} from '@/booking/application/commands/bookingConfirmationCommands.js';
import { GiftCardReference } from '@/booking/domain/primitive/giftCardReference.js';

describe('BookingCommands', () => {
  describe('PlaceBooking validation', () => {
    it('should fail if parameters are null', () => {
      const bookingId = generateId((value) => new BookingId(value));
      const amount = new Money(10);
      const giftCardReference = new GiftCardReference(Uuid.generate());

      expect(() => placeBooking(null as unknown as BookingId, amount, new Description('desc'), giftCardReference)).toThrow();
      expect(() => placeBooking(bookingId, null as unknown as Money, new Description('desc'), giftCardReference)).toThrow();
      expect(() => placeBooking(bookingId, amount, null as unknown as Description, giftCardReference)).toThrow();
      expect(() => placeBooking(bookingId, amount, new Description('desc'), null as unknown as GiftCardReference)).toThrow();
    });
  });

  describe('ConfirmBooking validation', () => {
    it('should fail if parameters are null', () => {
      const bookingId = generateId((value) => new BookingId(value));
      const giftCardReference = new GiftCardReference(Uuid.generate());
      const amount = new Money(10);

      expect(() => confirmBooking(null as unknown as BookingId, giftCardReference, amount)).toThrow();
      expect(() => confirmBooking(bookingId, null as unknown as GiftCardReference, amount)).toThrow();
      expect(() => confirmBooking(bookingId, giftCardReference, null as unknown as Money)).toThrow();
    });
  });

  describe('RejectBooking validation', () => {
    it('should fail if parameters are null', () => {
      const bookingId = generateId((value) => new BookingId(value));
      const giftCardReference = new GiftCardReference(Uuid.generate());
      const amount = new Money(10);

      expect(() => rejectBooking(null as unknown as BookingId, giftCardReference, amount)).toThrow();
      expect(() => rejectBooking(bookingId, null as unknown as GiftCardReference, amount)).toThrow();
      expect(() => rejectBooking(bookingId, giftCardReference, null as unknown as Money)).toThrow();
    });
  });
});
